config/module.go

```diff
@@ -233,72 +245,216 @@ const FirstRunSuccessSuffix = ".first_run_succeeded"
+		if err != nil {
+			return "", err
+		}
+		_, err = os.Stat(metaPath)
+		if err == nil {
+			// this is case 1, meta.json in exe dir
+			meta, err := parseJSONFile[JSONManifest](metaPath)
+			if err != nil {
+				return "", err
+			}
+			firstRun, err := utils.SafeJoinDir(unpackedModDir, meta.FirstRun)
+			if err != nil {
+				return "", err
+			}
+			firstRunPath, err := filepath.Abs(firstRun)
+			return firstRunPath, err
+		}
+	}
+	if m.NeedsSyntheticPackage() {
+		// this is case 2, side-by-side
+		// TODO(RSDK-7848): remove this case once java sdk supports internal meta.json.
+		metaPath, err := utils.SafeJoinDir(filepath.Dir(m.ExePath), "meta.json")
+		if err != nil {
+			return "", err
+		}
+		meta, err := parseJSONFile[JSONManifest](metaPath)
+		if err != nil {
+			// note: this error deprecates the side-by-side case because the side-by-side case is deprecated.
+			return "", errors.Wrapf(err, "couldn't find meta.json inside tarball %s (or next to it)", m.ExePath)
+		}
+		firstRun, err := utils.SafeJoinDir(unpackedModDir, meta.FirstRun)
+		if err != nil {
+			return "", err
+		}
+		firstRunPath, err := filepath.Abs(firstRun)
+		return firstRunPath, err
+	}
+	return "", errors.New("no first run script")
+}
+
+// FirstRun executes a module-specific setup script.
+func (m *Module) FirstRun(
+	ctx context.Context,
+	localPackagesDir,
+	dataDir string,
+	env map[string]string,
+	logger logging.Logger,
+) error {
+	logger = logger.Sublogger("first_run").WithFields("module", m.Name)
+
+	unpackedModDir, err := m.exeDir(localPackagesDir)
 	if err != nil {
-		return "", noop, err
+		return err
 	}
 
+	// check if FirstRun already ran successfully for this module version by
+	// checking if a success marker file exists on disk.
 	firstRunSuccessPath := unpackedModDir + FirstRunSuccessSuffix
 	if _, err := os.Stat(firstRunSuccessPath); !errors.Is(err, os.ErrNotExist) {
 		logger.Info("first run already ran")
-		return "", noop, errors.New("first run already ran")
+		return nil
 	}
-	markFirstRunSuccess := func() error {
-		//nolint:gosec // safe
-		_, err := os.Create(firstRunSuccessPath)
+
+	// Load the module's meta.json. If it doesn't exist or fails to load, debug log and exit quietly.
+	meta, err := m.getJSONManifest(unpackedModDir)
+	if err != nil {
+		logger.Debugw("failed to load meta.json, skipping setup phase", "error", err)
+		return nil
+	}
+	if meta.FirstRun == "" {
+		logger.Debug("no first run script specified, skipping setup phase")
+		return nil
+	}
+	relFirstRunPath, err := utils.SafeJoinDir(unpackedModDir, meta.FirstRun)
+	if err != nil {
+		logger.Errorw("failed to build path to first run script, skipping setup phase", "error", err)
+		return nil
+	}
+	firstRunPath, err := filepath.Abs(relFirstRunPath)
+	if err != nil {
+		logger.Errorw("failed to build absolute path to first run script, skipping setup phase", "path", relFirstRunPath, "error", err)
+		return nil
+	}
```

Comment:
these errors feel worthy of `ERROR` or `WARN` logs, since the user clearly specified a first run script, but it cannot be found.